src/frontend/src/optimizer/plan_node/stream_exchange.rs

```diff
@@ -43,13 +44,46 @@ impl StreamExchange {
             input.append_only(),
             input.watermark_columns().clone(),
         );
-        StreamExchange { base, input }
+        StreamExchange {
+            base,
+            input,
+            no_shuffle: false,
+        }
+    }
+
+    pub fn new_no_shuffle(input: PlanRef) -> Self {
+        let ctx = input.ctx();
+        let pk_indices = input.logical_pk().to_vec();
+        // Dispatch executor won't change the append-only behavior of the stream.
+        let base = PlanBase::new_stream(
+            ctx,
+            input.schema().clone(),
+            pk_indices,
+            input.functional_dependency().clone(),
+            input.distribution().clone(),
+            input.append_only(),
+            input.watermark_columns().clone(),
+        );
+        StreamExchange {
+            base,
+            input,
+            no_shuffle: true,
```

Comment:
When this flag is true, it seems more like a brand new kind of plan node. 🤔 Note that it's invalid for normal plans having such an exchange node.

Reply:
With this flag, we can express no shuffle edge in the stream plan node, otherwise, we need to handle it manually in the fragmenter.  I am wondering whether we can use it to represent the delta join in the stream plan.

Reply:
Should we use a brand-new exchange to separate the no-shuffle exchange from the normal one?🤔

Reply:
> I am wondering whether we can use it to represent the delta join in the stream plan.

Sounds exciting!

> Should we use a brand-new exchange to separate the no-shuffle exchange from the normal one?🤔

Either LGTM. Let's check which one is clearer once we have more usages in the future.